packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+  >;
+}
+
+/**
+ * Get the default state of the {@link PermissionController}.
+ *
+ * @template Permission - The controller's permission type union.
+ */
+function getDefaultState<Permission extends PermissionConstraint>() {
+  return { subjects: {} } as PermissionControllerState<Permission>;
+}
+
+/**
+ * Gets the state of the {@link PermissionController}.
+ */
+export type GetPermissionControllerState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => PermissionControllerState<PermissionConstraint>;
+};
+
+/**
+ * Gets the names of all subjects from the {@link PermissionController}.
+ */
+export type GetSubjects = {
+  type: `${typeof controllerName}:getSubjectNames`;
+  handler: () => (keyof PermissionControllerSubjects<PermissionConstraint>)[];
+};
+
+/**
+ * Checks whether the specified subject has any permissions.
+ */
+export type HasPermissions = {
+  type: `${typeof controllerName}:hasPermissions`;
+  handler: GenericPermissionController['hasPermissions'];
+};
+
+/**
+ * Clears all permissions from the {@link PermissionController}.
+ */
+export type ClearPermissions = {
+  type: `${typeof controllerName}:clearPermissions`;
+  handler: () => void;
+};
+
+// TODO: Implement all desired actions
+/**
+ * The {@link ControllerMessenger} actions of the {@link PermissionController}.
+ */
+export type PermissionControllerActions =
+  | ClearPermissions
+  | GetPermissionControllerState
+  | GetSubjects
+  | HasPermissions;
+
+/**
+ * The generic state change event of the {@link PermissionController}.
+ */
+export type PermissionControllerStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [PermissionControllerState<PermissionConstraint>, Patch[]];
+};
+
+/**
+ * The {@link ControllerMessenger} events of the {@link PermissionController}.
+ *
+ * The permission controller only emits its generic state change events.
+ * Consumers should use selector subscriptions to subscribe to relevant
+ * substate.
+ */
+export type PermissionControllerEvents = PermissionControllerStateChange;
+
+/**
+ * The external {@link ControllerMessenger} actions available to the
+ * {@link PermissionController}.
+ */
+type AllowedActions =
+  | AddApprovalRequest
+  | HasApprovalRequest
+  | AcceptApprovalRequest
+  | RejectApprovalRequest;
+
+/**
+ * The messenger of the {@link PermissionController}.
+ */
+export type PermissionControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  PermissionControllerActions | AllowedActions,
+  PermissionControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+/**
+ * A generic {@link PermissionController}.
+ */
+export type GenericPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+/**
+ * Options for {@link PermissionController.grantPermissions}.
+ */
+type GrantPermissionsOptions = {
+  approvedPermissions: RequestedPermissions;
+  subject: PermissionSubjectMetadata;
```

Comment:
Was there a future plan for this property to hold more than just the `origin`? Having to wrap the origin in an object and unwrap it again seems unnecessary.

`requestPermissions` for example just receives the `origin`. If there was additional metadata, isn't that one place it'd come in?

Reply:
`requestPermissions` now receives a `PermissionSubjectMetadata` object instead of just the `origin`: b4fae68

We decided to do this because we may soon want to pass in more subject metadata at that point in the near future, such as `isExtension` and `isSnap`.